simple/lint.go

```diff
@@ -1707,3 +1708,55 @@ func (c *Checker) LintRedundantSprintf(j *lint.Job) {
 		ast.Inspect(f, fn)
 	}
 }
+
+func (c *Checker) LintStringCopy(j *lint.Job) {
+	emptyStringLit := func(e ast.Expr) bool {
+		bl, ok := e.(*ast.BasicLit)
+		return ok && bl.Value == `""`
+	}
+	fn := func(node ast.Node) bool {
+		switch x := node.(type) {
+		case *ast.BinaryExpr: // "" + s, s + ""
+			if x.Op != token.ADD {
+				break
+			}
+			var want ast.Expr
+			switch {
+			case emptyStringLit(x.X):
+				want = x.Y
+			case emptyStringLit(x.Y):
+				want = x.X
+			default:
+				return true
+			}
+			j.Errorf(x, "should use %s instead of %s",
+				j.Render(want), j.Render(x))
+		case *ast.CallExpr: // string([]byte(s))
+			if len(x.Args) != 1 {
+				break
+			}
+			bt, ok := j.Program.Info.TypeOf(x.Fun).(*types.Basic)
+			if !ok || bt.Kind() != types.String {
+				break
+			}
+			nested, ok := x.Args[0].(*ast.CallExpr)
+			if !ok {
+				break
+			}
+			st, ok := j.Program.Info.TypeOf(nested.Fun).(*types.Slice)
+			if !ok {
+				break
+			}
+			et, ok := st.Elem().(*types.Basic)
```

Comment:
We need to check that `[]byte(x)` in `string([]byte(x))` is really converting a string to []byte. It could also be something like a json.RawMessage. Taken from the stdlib:

```
if string([]byte(*data.Id)) != raw {
	t.Fatalf("Raw mismatch: have %#q want %#q", []byte(*data.Id), raw)
}
```

where `data.Id` is a *json.RawMessage